server/server.go

```diff
@@ -190,22 +167,56 @@ func (sc *ServerConfig) LoadFromFile(file string) error {
-    serverCN := serverCertX509.Subject.CommonName
-    
-    if len(clientCN) == 0 {
-        return errors.New("The common name in the certificate is empty. The node ID must not be empty")
-    }
-     
-    if clientCN != serverCN {
-        return errors.New(fmt.Sprintf("Server and client certificates have differing common names(%s and %s). This is the string used to uniquely identify the node.", serverCN, clientCN))
+    if (YAMLTLSFiles{}) != ysc.TLS  {
+        rootCAs := x509.NewCertPool()
+        
+        if !rootCAs.AppendCertsFromPEM([]byte(ysc.TLS.RootCA)) {
+            return errors.New("Could not append root CA to chain")
+        }
+        
+        clientCertificate, _ := tls.X509KeyPair([]byte(ysc.TLS.ClientCertificate), []byte(ysc.TLS.ClientKey))
+        serverCertificate, _ := tls.X509KeyPair([]byte(ysc.TLS.ServerCertificate), []byte(ysc.TLS.ServerKey))
+        clientTLSConfig := &tls.Config{
+            Certificates: []tls.Certificate{ clientCertificate },
+            RootCAs: rootCAs,
+            GetClientCertificate: func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+                return &clientCertificate, nil
+            },
+        }
+        serverTLSConfig := &tls.Config{
+            Certificates: []tls.Certificate{ serverCertificate },
+            ClientCAs: rootCAs,
+        }
+        
+        sc.ServerTLS = serverTLSConfig
+        
+        
+        clientCertX509, _ := x509.ParseCertificate(clientCertificate.Certificate[0])
+        serverCertX509, _ := x509.ParseCertificate(serverCertificate.Certificate[0])
+        clientCN := clientCertX509.Subject.CommonName
+        serverCN := serverCertX509.Subject.CommonName
+        
+        if len(clientCN) == 0 {
+            return errors.New("The common name in the certificate is empty. The node ID must not be empty")
+        }
+         
+        if clientCN != serverCN {
+            return errors.New(fmt.Sprintf("Server and client certificates have differing common names(%s and %s). This is the string used to uniquely identify the node.", serverCN, clientCN))
+        }
+        if ysc.NodeID == ""  {
+                sc.NodeID = clientCN
+            } else {
+                sc.NodeID = ysc.NodeID
+            }
+            sc.Hub = NewHub(sc.NodeID, NewSyncController(uint(ysc.MaxSyncSessions), nil, ddbSync.NewPeriodicSyncScheduler(time.Millisecond * time.Duration(ysc.SyncSessionPeriod)), sc.SyncExplorationPathLimit), clientTLSConfig)
+    } else {
+        if ysc.NodeID == "" {
+            return errors.New("No TLS certificates provided. Config file must have 'nodeid' field")
+        }
+        sc.NodeID = ysc.NodeID
+        Log.Infof(" No TLS Config provided. Http mode\n")
+        sc.Hub = NewHub(sc.NodeID, NewSyncController(uint(ysc.MaxSyncSessions), nil, ddbSync.NewPeriodicSyncScheduler(time.Millisecond * time.Duration(ysc.SyncSessionPeriod)), sc.SyncExplorationPathLimit), nil)
```

Comment:
Again, just create this at one point in the code. Factor out similarities in conditional branch structures and do that stuff in one place.

Reply:
Got it.